diff --git a/libraries/qt-qrcodegenerator/CMakeLists.txt b/libraries/qt-qrcodegenerator/CMakeLists.txt
index e18da0e71..6a77c95d4 100644
--- a/libraries/qt-qrcodegenerator/CMakeLists.txt
+++ b/libraries/qt-qrcodegenerator/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.6)
+cmake_minimum_required(VERSION 3.10)

 project(qrcode)

@@ -18,15 +18,12 @@ elseif(Launcher_QT_VERSION_MAJOR EQUAL 6)
     list(APPEND systeminfo_LIBS Qt${QT_VERSION_MAJOR}::Core5Compat)
 endif()

-add_library(qrcode STATIC qr.h qr.cpp QR-Code-generator/cpp/qrcodegen.cpp QR-Code-generator/cpp/qrcodegen.hpp )
+find_package(qrcodegencpp REQUIRED)

-target_link_libraries(qrcode Qt${QT_VERSION_MAJOR}::Core Qt${QT_VERSION_MAJOR}::Gui ${systeminfo_LIBS})
+add_library(qrcode STATIC qr.h qr.cpp)

+target_include_directories(qrcode PUBLIC ./ PRIVATE ${qrcodegencpp_INCLUDE_DIR})
+target_link_libraries(qrcode Qt${QT_VERSION_MAJOR}::Core Qt${QT_VERSION_MAJOR}::Gui ${systeminfo_LIBS} qrcodegencpp::qrcodegencpp)

 # needed for statically linked qrcode in shared libs on x86_64
-set_target_properties(qrcode
-    PROPERTIES POSITION_INDEPENDENT_CODE TRUE
-)
-
-target_include_directories(qrcode PUBLIC ./ PRIVATE QR-Code-generator/cpp/)
-
+set_target_properties(qrcode PROPERTIES POSITION_INDEPENDENT_CODE TRUE)
